### PR TITLE
provider/azurerm: Add resources virtual network gateway and connection

### PR DESCRIPTION
**Pull request by dominik-lekse**
This pull request adds the resources `azurerm_virtual_network_gateway` and `azurerm_virtual_network_gateway_connection` which can be used to manage [Azure VPN Gateways and Connections](https://docs.microsoft.com/en-us/azure/vpn-gateway/vpn-gateway-about-vpngateways).

In pull request https://github.com/hashicorp/terraform/pull/9255, @pmcatominey provided the major part the resource `azurerm_virtual_network_gateway` implementation. This pull request bases on the top of this and contains the original commits of https://github.com/hashicorp/terraform/pull/9255. In addition, the resource `azurerm_virtual_network_gateway` has been completed and implementation and tests of the resource `azurerm_virtual_network_gateway_connection` have been added. Further, documentation for both resources is included.

Please consider the following notes when reviewing:

- Expect the resource `azurerm_virtual_network_gateway` to run for a while since "creating a gateway can take a long time to complete. Often 45 minutes or more." (see https://docs.microsoft.com/en-us/azure/vpn-gateway/vpn-gateway-create-site-to-site-rm-powershell#a-namecreategatewaya6-create-the-virtual-network-gateway). The Terraform resource polls for the status until the virtual network gateways is fully provisioned and can be used with other resources such as connections.
- In https://github.com/hashicorp/terraform/pull/9255, @pmcatominey mentioned a bug of the Azure management API when destroying a gateway subnet which has been associated with a virtual network gateway which has been successfully destroyed seconds before. I can confirm that the bug is still present, i.e. that such a gateway subnet cannot be deleted for a couple of minutes. Unfortunately, the API does not expose a status on such a subnet which we could use to distinguish. Instead of waiting for a fixed length of time at the end of destroying a virtual network gateway, I have implemented an alternative solution. When a gateway subnet cannot be deleted with a response which occurs when an associated virtual network gateway has been removed previously, Terraform will retry the deletion for a fixed period of time. In my opinion, this workaround is reasonable although it has to be implemented in the subnet resource.
- Both resources should support Azure ExpressRoute functionality, however this is not part of the test coverage since I do not have access to an Azure subscription which is ready for this.

Acceptance tests:

- `TestAccAzureRMVirtualNetworkGateway_basic`: Create a virtual network gateway and destroys it
- `TestAccAzureRMVirtualNetworkGatewayConnection_sitetosite`: Creates a site-to-site connection and destroys it
- `TestAccAzureRMVirtualNetworkGatewayConnection_vnettovnet`: Creates virtual network gateways in two Azure regions and connects them using a vnet-to-vnet connection
- `TestAccAzureRMVirtualNetworkGateway_importBasic`: Verifies import of a virtual network gateway
- `TestAccAzureRMVirtualNetworkGatewayConnection_importSiteToSite`: Verifies import of a virtual network gateway connection

```
=== RUN   TestAccAzureRMVirtualNetworkGateway_basic
--- PASS: TestAccAzureRMVirtualNetworkGateway_basic (2190.70s)
=== RUN   TestAccAzureRMVirtualNetworkGateway_importBasic
--- PASS: TestAccAzureRMVirtualNetworkGateway_importBasic (2280.54s)
=== RUN   TestAccAzureRMVirtualNetworkGatewayConnection_sitetosite
--- PASS: TestAccAzureRMVirtualNetworkGatewayConnection_sitetosite (2401.24s)
=== RUN   TestAccAzureRMVirtualNetworkGatewayConnection_importSiteToSite
--- PASS: TestAccAzureRMVirtualNetworkGatewayConnection_importSiteToSite (2664.22s)
=== RUN   TestAccAzureRMVirtualNetworkGatewayConnection_vnettovnet
--- PASS: TestAccAzureRMVirtualNetworkGatewayConnection_vnettovnet (2480.51s)
```

Open tasks:

- [x] Successful create and delete resources using `azurerm_virtual_network_gateway` and `azurerm_virtual_network_gateway_connection` in a scenario
- [x] Implement test for `azurerm_virtual_network_gateway` (`TestAccAzureRMVirtualNetworkGateway_basic`)
- [x] Implement import test for `azurerm_virtual_network_gateway` (`TestAccAzureRMVirtualNetworkGateway_importBasic`)
- [x] Documentation for `azurerm_virtual_network_gateway`
- [x] Implement test for `azurerm_virtual_network_gateway_connection` to test site-to-site connection (`TestAccAzureRMVirtualNetworkGatewayConnection_sitetosite`)
- [x] Implement test for `azurerm_virtual_network_gateway_connection` to test vnet-to-vnet connection (`TestAccAzureRMVirtualNetworkGatewayConnection_vnettovnet`)
- [x] Implement import test for `azurerm_virtual_network_gateway_connection` (`TestAccAzureRMVirtualNetworkGatewayConnection_importSiteToSite`)
- [x] Documentation for `azurerm_virtual_network_gateway_connection`
- [ ] Review

This pull request will resolve the issues, https://github.com/hashicorp/terraform/issues/8372, https://github.com/hashicorp/terraform/issues/10634 and includes the pull request https://github.com/hashicorp/terraform/pull/9255. It would be great to get feedback from you waiting for these resources.

// Dominik

### CODE REVIEW

**Comment by anniehedgpeth** [comment]
Hey @tombuildsstuff, any word on when this will be merged? I have a new template that requires this resource. :) Thanks!

**Comment by dominik-lekse** [comment]
Just resolved recent conflicts with master.

I assume the Terraform team has fixed their sprint goal for the current release cycle and there are also lots of other community pull requests which consume review time.

Would appreciate if this could fit in one of the upcoming release cycles.

**Comment by dominik-lekse** [comment]
Thanks for reviewing @tombuildsstuff. I will be able to go through your points until the end of the week.

**Comment by YuriyKischenko** [comment]
Ping @dominik-lekse

**Comment by dominik-lekse** [comment]
Sorry for being silent here for a while. Unfortunately, working on this pull request fall behind other priorities.

I will try to catch up on this Friday with the comments.

In particular, I am sure a rebase with the current master is necessary since #14004 is merged.

As a request to the community: We can also collaborate on finishing this pull request. To those of you eager to collaborate: Just ping me here and I invite you with write permissions to the fork which is the basis for this pull request https://github.com/dominik-lekse/terraform/tree/public-feature/azurerm-vpn-gateway.

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+			"routing_weight": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  10,
+			},
+
+			"shared_key": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Sensitive: true,
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go]: The shared key needs to be provided as the API does not generate one. I suggest to keep this as it is.

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+			},
+
+			"connection_status": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go]: You are right, the connection_status cannot be set, it is read-only. The idea to keep this was to allow the user to use it as part of an output variable. However, in this case, a separate data source is the better option right?

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+					string(network.VirtualNetworkGatewayConnectionStatusConnected),
+					string(network.VirtualNetworkGatewayConnectionStatusConnecting),
+					string(network.VirtualNetworkGatewayConnectionStatusNotConnected),
+					string(network.VirtualNetworkGatewayConnectionStatusUnknown),
+				}, false),
+			},
+
+			"egress_bytes_transferred": {
+				Type:     schema.TypeInt,
+				Optional: true,
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go]: See above

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+				Computed: true,
+			},
+
+			"ingress_bytes_transferred": {
+				Type:     schema.TypeInt,
+				Optional: true,
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go]: See above

**Comment by dominik-lekse** [comment]
I will reopen the pull request in the new azurerm provider repository with the requested changes.

**Comment by tombuildsstuff** [comment]
That'd be awesome - thanks @dominik-lekse :)

**Comment by YuriyKischenko** [comment]
Hi @dominik-lekse 
Microsoft have just introduced new SKUs for VPN: https://azure.microsoft.com/en-us/blog/new-azure-vpn-gateways-now-6x-faster/

Could you please update your pull request

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway.go [hunk 5]
```diff
@@ -0,0 +1,685 @@
...
+				ValidateFunc: validation.StringInSlice([]string{
+					string(network.RouteBased),
+					string(network.PolicyBased),
+				}, false),
+				ForceNew: true,
+			},
+
+			"enable_bgp": {
+				Type:     schema.TypeBool,
+				Optional: true,
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway.go]: Done.

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway.go [hunk 6]
```diff
@@ -0,0 +1,685 @@
...
+			},
+
+			"active_active": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway.go]: Done.

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway.go [hunk 7]
```diff
@@ -0,0 +1,685 @@
...
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								string(network.VirtualNetworkGatewaySkuNameBasic),
+								string(network.VirtualNetworkGatewaySkuNameStandard),
+								string(network.VirtualNetworkGatewaySkuNameHighPerformance),
+								string(network.VirtualNetworkGatewaySkuNameUltraPerformance),
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway.go]: As an alternative, I would add a validation function which cross checks gateway type and sku. Would that be fine?

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway.go [hunk 8]
```diff
@@ -0,0 +1,685 @@
...
+							}, false),
+						},
+						"tier": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								string(network.VirtualNetworkGatewaySkuTierBasic),
+								string(network.VirtualNetworkGatewaySkuTierStandard),
+								string(network.VirtualNetworkGatewaySkuTierHighPerformance),
+								string(network.VirtualNetworkGatewaySkuTierUltraPerformance),
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway.go]: Done.

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway.go [hunk 9]
```diff
@@ -0,0 +1,685 @@
...
+						},
+					},
+				},
+				Set: hashVirtualNetworkGatewaySku,
+			},
+
+			"ip_configuration": {
+				Type:     schema.TypeList,
+				Required: true,
+				MinItems: 1,
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway.go]: I do not see how we can remove the required for the ip_configuration block. At least one is required to associate a public ip address (refer to https://docs.microsoft.com/en-us/cli/azure/network/vnet-gateway#create).

However, maybe I misunderstood your point.

Comment (tombuildsstuff) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway.go]: @dominik-lekse sorry, to be clearer - `MinItems: 1` can be inferred as `Required` is set to `true` - thus we can remove this line :)

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway.go [hunk 10]
```diff
@@ -0,0 +1,685 @@
...
+				MaxItems: 2,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "vnetGatewayConfig",
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway.go]: In this case, we would need to change this field to a required one. This name is an internal name but required by the API. I.e. the API does not apply a default name if this is not provided. I tried to replicate the function of Azure CLI or the Portal which which uses "vnetGatewayConfig" as the value for this field.

For this reason, I suggest to keep the default value.

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway.go (hunk 9 above)
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway.go]: Alright, done.

File: builtin/providers/azurerm/resource_arm_local_network_gateway.go [hunk 11]
```diff
@@ -147,3 +141,33 @@ func resourceArmLocalNetworkGatewayDelete(d *schema.ResourceData, meta interface
...
+	if err != nil {
+		return "", "", err
+	}
+	name := id.Path["localNetworkGateways"]
+	resGroup := id.ResourceGroup
+
+	return resGroup, name, nil
+}
+
+func retrieveLocalNetworkGatewayById(localNetworkGatewayId string, meta interface{}) (*network.LocalNetworkGateway, bool, error) {
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_local_network_gateway.go]: Done.

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway.go [hunk 12]
```diff
@@ -0,0 +1,685 @@
...
+
+	// Gateways are not fully cleaned up when the API indicates the delete operation
+	// has finished, this workaround was suggested by Azure support to avoid conflicts
+	// when modifying/deleting the related subnet or network. Although the API indicated
+	// that the Virtual Network Gateway does not exist anymore, there is still a link
+	// to the Gateway Subnet it has been associated with. This causes an error when
+	// trying to delete the Gateway Subnet immediately after the Virtual Network Gateway
+	// has been deleted.
+
+	d.SetId("")
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway.go]: Regarding this issue, I tried some of the alternatives which are definitly better approaches before "relying" on retries of subnet deletion attempts.

Regaring the approach to "poll for the deletion of the Virtual Network Gateway using a StateFunc and a ContinuousTargetOccurence value":  I tried exactly this in the gateway resource, but due to the consistency bug the API responds with a 400 error, as if the gateway would have never existed. In this case, the "hidden link" between the gateway subnet and the gateway still exists.

I tried polling using vnetGatewayClient.Get only. I will check the other methods you metioned before: "see if the Virtual Network Gateway is visible through the either Virtual Network Gateway's List method / Resource Group's contents".

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+	"log"
+	"net/http"
+	"time"
+)
+
+func resourceArmVirtualNetworkGatewayConnection() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmVirtualNetworkGatewayConnectionCreate,
+		Read:   resourceArmVirtualNetworkGatewayConnectionRead,
+		Update: resourceArmVirtualNetworkGatewayConnectionCreate,
```
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go]: Done.

File: builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go (hunk 2 above)
Comment (dominik-lekse) [inline builtin/providers/azurerm/resource_arm_virtual_network_gateway_connection.go]: I will move this to a separate data source.

**Comment by dominik-lekse** [comment]
@YuriyKischenko I have implemented the new SKUs, they will be part of the pull request I will open in the https://github.com/terraform-providers/terraform-provider-azurerm repository soon.

Even with the new gateway types, running the tests still takes > 2 hours :(

**Comment by tombuildsstuff** [comment]
Closing the original PR since this has been migrated to the new repo: https://github.com/terraform-providers/terraform-provider-azurerm/pull/133

**Pull request closed without merging**